pkgs/top-level/named-hierarchy-files.nix

```diff
@@ -0,0 +1,108 @@
+/*
+Internal Nixpkgs file, it may be removed or changed at any point without warning
+
+This file contains the mapping from attribute name to the file in pkgs/by-name that contains a definition of it.
+No validity checks are performed, instead this file is optimized for performance.
+TODO: Link to manual and validity checks
+
+Example: If `pkgs/by-name/he/hello` exists, the result will have a `hello = pkgs/by-name/he/hello/package.nix` entry.
+
+Type: AttrsOf Path
+*/
+
+# Because the result might be fairly large, we only want compute it once.
+# In order to ensure this, we make use of Nix's `import` cache.
+# This cache memoizes the resulting value of the imported file, notably including any `let in` bindings at the top of the file.
+# This is why everything here is computed in a `let in` binding at the top.
+let
+
+  # We don't _need_ `lib` here, but importing it wouldn't be detrimental to performance either because of Nix's `import` cache
+  inherit (builtins) attrNames elemAt length mapAttrs readDir;
+
+  # Warning: Do not make this configurable, otherwise the `import` evaluation cache won't work!
+  baseDirectory = ../by-name;
```

Comment:
It shouldn't be in `lib` (yet) because it's only an internal standard for now. The suggested `isAttrs` test should take care of ensuring this, I'll do that.

Reply:
That's fair. A separate file may already be helpful for testing and achieve some of the same effect.

Reply:
`binaryMerge` could be added to `lib` before anything `callPackage`-specific.

Reply:
Sounds good, though note that this will delay this PR further.

I'll go for an interface like this (name bikeshedding appreciated):
`mapMergeAttrs :: (Any -> Attrs) -> [ Any ] -> Attrs`

this should work with minimal performance loss

Reply:
I opened https://github.com/NixOS/nixpkgs/pull/244187 for this

Though I changed the function to just be `lib.attrsets.mergeAttrsList :: [ Attrs ] -> Attrs`. I measured it, there's no performance loss with this.